engine/src/xcc.rs

```diff
@@ -0,0 +1,172 @@
+pub const VERSION_KEY: &[u8] = b"version";
+pub const CODE_KEY: &[u8] = b"router_code";
+pub const VERSION_UPDATE_GAS: NearGas = NearGas::new(5_000_000_000_000);
+pub const INITIALIZE_GAS: NearGas = NearGas::new(5_000_000_000_000);
+/// Amount of NEAR needed to cover storage for a router contract.
+pub const STORAGE_AMOUNT: Yocto = Yocto::new(5_000_000_000_000_000_000_000_000);
+
+/// Type wrapper for version of router contracts.
+#[derive(
+    Debug, Clone, Copy, Default, PartialEq, Eq, PartialOrd, Ord, BorshDeserialize, BorshSerialize,
+)]
+pub struct CodeVersion(pub u32);
+
+impl CodeVersion {
+    pub fn increment(self) -> Self {
+        Self(self.0 + 1)
+    }
+}
+
+/// Type wrapper for router bytecode.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct RouterCode(pub Vec<u8>);
+
+#[derive(Debug, Clone, PartialEq, Eq, BorshDeserialize, BorshSerialize)]
+pub struct AddressVersionUpdateArgs {
+    pub address: Address,
+    pub version: CodeVersion,
+}
+
+pub fn handle_precomile_promise<I, P>(
+    io: &I,
+    handler: &mut P,
+    promise: PromiseCreateArgs,
+    current_account_id: &AccountId,
+) where
+    P: PromiseHandler,
+    I: IO,
+{
+    let target_account: &str = promise.target_account_id.as_ref();
+    let sender = Address::decode(&target_account[0..40]).expect(ERR_INVALID_ACCOUNT);
+    let latest_code_version = get_latest_code_version(io);
+    let sender_code_version = get_code_version_of_address(io, &sender);
+    let deploy_needed = match sender_code_version {
+        None => AddressVersionStatus::DeployNeeded {
+            create_needed: true,
+        },
+        Some(version) if version < latest_code_version => AddressVersionStatus::DeployNeeded {
+            create_needed: false,
+        },
+        Some(_version) => AddressVersionStatus::UpToDate,
+    };
+    let _promise_id = match deploy_needed {
+        AddressVersionStatus::DeployNeeded { create_needed } => {
+            let mut promise_actions = Vec::with_capacity(4);
+            if create_needed {
+                promise_actions.push(PromiseAction::CreateAccount);
+                promise_actions.push(PromiseAction::Transfer {
+                    amount: STORAGE_AMOUNT,
+                });
+            }
+            promise_actions.push(PromiseAction::DeployConotract {
+                code: get_router_code(io).0,
+            });
+            // After a deploy we call the contract's initialize function
+            promise_actions.push(PromiseAction::FunctionCall {
+                name: "initialize".into(),
+                args: Vec::new(),
+                attached_yocto: ZERO_YOCTO,
+                gas: INITIALIZE_GAS,
+            });
+            // After the contract is deployed and initialized, we can call the method requested
+            promise_actions.push(PromiseAction::FunctionCall {
+                name: promise.method,
+                args: promise.args,
+                attached_yocto: promise.attached_balance,
```

Comment:
I see now that in the precompile we always use zero as the attached balance. Anyway, I still think we should add here a check that this amount is zero since we are not expecting any value greater than zero now or in the future. At least not without extra logic that makes the user pay for this balance in advance.

Reply:
Added a check promise.attached_balance is zero in https://github.com/aurora-is-near/aurora-engine/pull/560/commits/0b55a183fc7fd5f3b8435bf8c09a128b49d5d4cd